subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorNestingIntegrationTest.groovy

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers.internal
+
+import spock.lang.Unroll
+
+class WorkerExecutorNestingIntegrationTest extends AbstractWorkerExecutorIntegrationTest {
+
+    @Unroll
+    def "workers with no isolation can spawn more work with #nestedIsolationMode"() {
+        buildFile << """
+            ${getRunnableWithNesting("IsolationMode.NONE", nestedIsolationMode)}
+            task runInWorker(type: NestingWorkerTask)
+        """.stripIndent()
+
+        when:
+        def gradle = executer.withTasks("runInWorker").start()
+
+        then:
+        gradle.waitForFinish()
+
+        and:
+        gradle.standardOutput.contains("Hello World")
+
+        where:
+        nestedIsolationMode << ISOLATION_MODES
+    }
+
+    /*
+     * This is not intended, but current behavior. We'll need to find a way to pass the service
+     * registry across the classloader isolation barrier.
+     */
+    @Unroll
+    def "workers with classpath isolation cannot spawn more work with #nestedIsolationMode"() {
+        buildFile << """
+            ${getRunnableWithNesting("IsolationMode.CLASSPATH", nestedIsolationMode)}
+            task runInWorker(type: NestingWorkerTask)
+        """.stripIndent()
+
+        when:
+        def gradle = executer.withTasks("runInWorker").start()
+
+        then:
+        gradle.waitForFailure()
```

Comment:
Should probably just use `fails()` here, too.  And check for specific error to avoid masking failures for other reasons.